diff -u --recursive ../build/clean/linux-5.4.117/drivers/Kconfig linux-5.4.117/drivers/Kconfig
--- ../build/clean/linux-5.4.117/drivers/Kconfig	2021-05-07 08:51:38.000000000 +0000
+++ linux-5.4.117/drivers/Kconfig	2022-02-16 14:03:39.241112549 +0000
@@ -9,6 +9,7 @@
 source "drivers/pcmcia/Kconfig"
 source "drivers/rapidio/Kconfig"
 
+source "drivers/uefidev/Kconfig"
 
 source "drivers/base/Kconfig"
 
diff -u --recursive ../build/clean/linux-5.4.117/drivers/Makefile linux-5.4.117/drivers/Makefile
--- ../build/clean/linux-5.4.117/drivers/Makefile	2021-05-07 08:51:38.000000000 +0000
+++ linux-5.4.117/drivers/Makefile	2022-02-16 14:05:22.684515177 +0000
@@ -186,3 +186,5 @@
 obj-$(CONFIG_GNSS)		+= gnss/
 obj-$(CONFIG_INTERCONNECT)	+= interconnect/
 obj-$(CONFIG_COUNTER)		+= counter/
+
+obj-$(CONFIG_UEFIDEV)		+= uefidev/
